llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -136,13 +140,115 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorShiftsByZero(
   return true;
 }
 
+bool AIE2PreLegalizerCombinerImpl::tryToCombineSetExtract(
+    MachineInstr &MI) const {
+  const Register DstReg = MI.getOperand(0).getReg();
+  MachineInstr *ExtOp = getDefIgnoringCopies(MI.getOperand(2).getReg(), MRI);
+
+  if (!isa<GIntrinsic>(MI) || !isa<GIntrinsic>(*ExtOp))
+    return false;
+  switch (cast<GIntrinsic>(MI).getIntrinsicID()) {
+  case Intrinsic::aie2_set_I512_I128: {
+    if (cast<GIntrinsic>(*ExtOp).getIntrinsicID() !=
+        Intrinsic::aie2_extract_I128_I512)
+      return false;
+    break;
+  }
+  case Intrinsic::aie2_set_I512_I256: {
+    const Register IdxReg = MI.getOperand(3).getReg();
+    if (cast<GIntrinsic>(*ExtOp).getIntrinsicID() !=
+            Intrinsic::aie2_ext_I256_I512 ||
+        IdxReg != ExtOp->getOperand(3).getReg())
+      return false;
+    break;
+  }
+  default:
+    return false;
+  }
+
+  MachineIRBuilder MIRBuilder(MI);
+  MIRBuilder.buildCopy(DstReg, ExtOp->getOperand(2).getReg());
+  MI.eraseFromParent();
+
+  return true;
+}
+
+std::optional<std::pair<std::map<int, Register>, int>>
```

Comment:
Can we have some comment about this complex return type?

Reply:
I think when I suggested outlining that code, I imagined a `std::map<int, Register>` return type. Is there anything that prevents using this simpler return type @abhinay-anubola ?

Reply:
While building `G_BUILD_VETCOR_TRUNC` we need the size of vector, size is calculated in this function and returned.

Reply:
Here's what I would do: I think this function should only return something when it could safely follow the chain of VINSERT. This means:
 - We ended up on a `set` intrinsic (so far I think it's sufficient to consider it the only safe "source")
 - We collected the right amount of lanes to cover the size of the input vector to the `set` intrinsic

In that case, we return a map with all the lane mappings, and otherwise we just return an empty map.

We then do not need to return the "source vector" size, because then in `tryToCombineVectorInserts` we can just iterate on every lane from `0` to `InsertIndices.size()`